Sources/DipError.swift

```diff
@@ -89,6 +98,87 @@ public enum DipError: Error, CustomStringConvertible {
       return "Resolved instance \(resolved ?? "nil") does not implement expected type \(key.type)."
     }
   }
+}
+
+extension DipError {
   
+  /// Informs you if the resolve failed because recursive limit was reached. Warning: O(n) for depth. May be slow
+  public var isRecusiveError: Bool {
+    switch self {
+      case let .autoWiringFailed(type: _, underlyingError: underlyingError):
+        if let dipError = underlyingError as? DipError {
+          return dipError.isRecusiveError
+        }
+        return false
+      case .recursionDepthReached:
+        return true
+      default:
+        return false
+    }
+  }
+  
+  public struct RecuriveErrorReport: CustomStringConvertible {
+    //The enture resolve stack that resulted in the recursive depth error
+    public let resolveStack: [Any.Type]
+    
+    //The sub stack that was identified as a cycle, if any.
+    public let cycleSubStack: [Any.Type]
+    
+    public var description: String {
+      if(cycleSubStack.count > 0) {
+        return "Recursive Depth exceeded: Cycle Found: \(cycleSubStack)"
+      }
+      return "Recursive Depth exceeded. No cycle found (inconclusive). Full stack: \(resolveStack)"
+    }
+  }
+  
+  public func analyzeRecursiveError() -> RecuriveErrorReport? {
+    if !isRecusiveError {
+      return nil
+    }
+    
+    var resolveStack = [Any.Type]()
+    
+    func buildStack(error: DipError) {
+      switch error {
+        case let .autoWiringFailed(type: type, underlyingError: underlyingError):
+          resolveStack.append(type)
+          if let dipError = underlyingError as? DipError {
+              buildStack(error: dipError)
+          }
+          return
+        case .recursionDepthReached:
+          return
+        default:
+          return
+      }
+    }
+    
+    buildStack(error: self)
+    
+    //Search for cycle subStack
+    var cycleSubStack = [Any.Type]()
+    var hitCount = [String:Int]()
```

Comment:
styling:
```swift
var cycleSubStack: [Any.Type] = []
var hitCount: [String:Int] = []
```

Reply:
?? I'm not aware of us having a preference on this one. 
Its not like we do 
`let foo: Foo = Foo()` either...

Reply:
well, it's confusing but we do reference two style guides in our docs unfortunately. One of them is more kept up to date and more comprehensive. That one says that for arrays / dictionaries, we should inform the type and initialize with the short form, which is either `[]` or `[:]`

ref: https://github.com/ck-private/mob_ios/blob/stable/Documentation/Styleguide/STYLEGUIDE.md